### DISCUSSION TITLE
"$raw.DiffuseColor|uvtrafo" material property value changes after FBX export

### DISCUSSION THREAD

**Discussion by Pankaj003** [assimp/assimp#3956] [Q&A]
Hi,
I am facing an issue with FBX export file. 
I am importing a fbx file and after importing, aiScene* is having below material property
< MatProperty key="$raw.DiffuseColor|uvtrafo" 
	type="float" tex_usage="Unknown" tex_index="0" size="5">
	0.500000 0.500000 0.152366 0.153189 0.000000 
< /MatProperty >

I am exporting the same aiScene* without and modification and on importing it back the property is getting changed to 
			< MatProperty key="$raw.DiffuseColor|uvtrafo" 
			type="float" tex_usage="Unknown" tex_index="0" size="5">
				0.000000 0.000000 1.000000 1.000000 0.000000 
			< /MatProperty >

This is causing uv transformation issue in my program. I am not using any Preprocessing flag while exporting the scene.
Any suggestion on how to fix this uvtrafo change? Is Assimp is doing any optimization while exporting the scene?

Also, I've observed that MatPropertyList number is reduced on exporting the fbx file.

Attached xml dump files before and after exporting.
[Dump.zip](https://github.com/assimp/assimp/files/6641725/Dump.zip)

Thanks

**Comment by kimkulling** [comment]
This seems to be some kind of a default material definition (0.0 and 1.0 looks like an initializer value). Are they any other materials in the FBX file?

**Comment by Pankaj003** [comment]
Hi @kimkulling,

Thanks for looking into the issue.
There's no other materials. I am sharing the FBX files (binary and ascii format), original and same exported using Assimp export call. 
You can compare the ASCII files and will observe that there are lot of difference in both of them. Primarily I am concerned with uv transformations.

FBX files:
[FBXFiles.zip](https://github.com/assimp/assimp/files/6667426/FBXFiles.zip)

Thanks,

**Comment by Pankaj003** [comment]
Hi @kimkulling,

I debugged the Assimp library and found the issue in void FBXExporter::WriteObjects () function.  I made the changes in the call and it fixed my issue.
**Below is my observation:**
**For the below block:**
	Texture: 1000006, "Texture::", "" {
		Type: "TextureVideoClip"
		Version: 202
		TextureName: "Texture::"
		Properties70:  {
			P: "CurrentTextureBlendMode", "enum", "", "", 0
			P: "UseMaterial", "bool", "", "", 1
		}

 **Properties70 block is not writing the translation and rotation information.**

**In the function void FBXExporter::WriteObjects () (file code\FBX\FBXExporter.cpp)**
  we have below implementation (line 1615 - 1628)
/////////////////////////////////////////////////////////////////////////////
///Code snippet from Assimp library call FBXExporter::WriteObjects ()
          // now write the actual texture node
            FBX::Node tnode("Texture");
            // TODO: some way to determine texture name?
            const std::string texture_name = "" + FBX::SEPARATOR + "Texture";
            tnode.AddProperties(texture_uid, texture_name, "");
            // there really doesn't seem to be a better type than this:
            tnode.AddChild("Type", "TextureVideoClip");
            tnode.AddChild("Version", int32_t(202));
            tnode.AddChild("TextureName", texture_name);
            FBX::Node p("Properties70");
            p.AddP70enum("CurrentTextureBlendMode", 0); // TODO: verify
            //p.AddP70string("UVSet", ""); // TODO: how should this work?
            p.AddP70bool("UseMaterial", 1);
            tnode.AddChild(p);
/////////////////////////////////////////////////////////////////////////////////////////

**FBX::Node p("Properties70"); is not writing transformation data, I modified this function block as below and it  fixed the issue**

///////////////////////////////////////////////////////////////////////////////////////////
///Modified block
          ///New code begins
           aiUVTransform trafo;
            unsigned int max = sizeof(aiUVTransform);
            aiGetMaterialFloatArray(mat, AI_MATKEY_UVTRANSFORM(aiTextureType_DIFFUSE, 0), (float*)&trafo, &max);
         ///New code ends
// now write the actual texture node
            FBX::Node tnode("Texture");
            // TODO: some way to determine texture name?
            const std::string texture_name = "" + FBX::SEPARATOR + "Texture";
            tnode.AddProperties(texture_uid, texture_name, "");
            // there really doesn't seem to be a better type than this:
            tnode.AddChild("Type", "TextureVideoClip");
            tnode.AddChild("Version", int32_t(202));
            tnode.AddChild("TextureName", texture_name);
            FBX::Node p("Properties70");
///New code begins
            p.AddP70vectorA("Translation", trafo.mTranslation[0], trafo.mTranslation[1], 0.0);
            p.AddP70vectorA("Scaling", trafo.mScaling[0], trafo.mScaling[1], 0.0);
 ///New code ends
            p.AddP70enum("CurrentTextureBlendMode", 0); // TODO: verify
            //p.AddP70string("UVSet", ""); // TODO: how should this work?
            p.AddP70bool("UseMaterial", 1);
            tnode.AddChild(p);
//////////////////////////////////////////////////////////////////////////////////////////////////////////////

Please have a look and let me know if my observations are correct and are we missing any other information here.
Attached the modified file for your reference.

[FBXExporter.zip](https://github.com/assimp/assimp/files/6681879/FBXExporter.zip)

Thanks,
Pankaj

**Comment by Pankaj003** [comment]
Hi @kimkulling,

I've changed few more files to add support for texture rotation read and write.
Please have a look and review them.

[AssimpChanges.zip](https://github.com/assimp/assimp/files/6687012/AssimpChanges.zip)

Thanks,
Pankaj
